compiler/sem/sanitizer.nim

```diff
@@ -0,0 +1,976 @@
+    checkSonsLen(n, 2)
+    result = prepareFrom(n)
+    result[0] = qualifiedIdent(c, n[0])
+    # note: we also allow normal string literals here (the parser doesn't)
+    result[1] = parseStrLit(c, n[1])
+  of nkInfix:
+    checkSonsLen(n, 3)
+    result = prepareFrom(n)
+    result[0] = qualifiedIdent(c, n[0])
+    result[1] = expr(c, n[1])
+    result[2] = expr(c, n[2])
+  of nkPrefix:
+    checkSonsLen(n, 2)
+    result = prepareFrom(n)
+    result[0] = qualifiedIdent(c, n[0])
+    result[1] = expr(c, n[1])
+  of nkPostfix:
+    invalid()
+  of nkHiddenCallConv:
+    # turn it into a normal call
+    checkMinSonsLen(n, 1)
+    result = newNodeI(nkCall, n.info, n.len)
+    for i in 0..<n.len:
+      result[i] = processArg(c, n[i])
+
+  of nkExprEqExpr, nkExprColonExpr, nkIdentDefs, nkVarTuple:
+    # these are only allowed in certain contexts
+    invalid()
+  of nkPar:
+    # an ``nkPar`` as an expression must only have a single child
+    checkSonsLen(n, 1)
+    result = newTreeI(nkPar, n.info): expr(c, n[1])
+  of nkObjConstr:
+    checkMinSonsLen(n, 1)
+    result = newNodeI(nkObjConstr, n.info, n.len)
+    result[0] = parseTypeNode(c, n[0])
+    for i in 1..<n.len:
+      let it = n[i]
+      if it.kind == nkExprColonExpr:
+        result[i] = newTreeI(nkExprEqExpr, it.info): [parseIdent(c, it[0]), expr(c, it[1])]
+      else:
+        result[i] = invalidAst(c, it)
+
+  of nkCurly:
+    result = exprs(c, n)
+  of nkCurlyExpr:
+    checkMinSonsLen(n, 1)
+    result = exprs(c, n)
+  of nkBracket:
+    result = exprs(c, n)
+  of nkBracketExpr:
+    # only a single sub-node is required, because ``nkBracketExpr`` is also used
+    # for ``n[]``
+    checkMinSonsLen(n, 1)
+    result = exprs(c, n)
+  of nkPragmaExpr:
+    unreachable("missing")
+  of nkRange:
+    checkSonsLen(n, 2)
+    result = newTreeI(nkInfix, n.info, ident(c, ".."), expr(c, n[0]), expr(c, n[1]))
+  of nkDotExpr:
+    result = parseDotExpr(c, n)
+  of nkCheckedFieldExpr:
+    # tricky. We *could* emit an 'if' statement from the condition with a
+    # raise statement inside, but that might be a bit unintuitive and would
+    # also mean that the processing doesn't roundtrip when sem-checking the
+    # result and then passing it back into the sanitizer. Instead, we discard
+    # the check part
+    checkSonsLen(n, 2)
+    if n[0].kind == nkDotExpr:
+      result = parseDotExpr(c, n[0])
+    else:
+      invalid()
+  of nkDerefExpr:
+    checkSonsLen(n, 1)
+    result = newTreeI(nkDerefExpr, n.info, toUntyped n[0])
+  of nkIfExpr, nkIfStmt, nkWhenStmt:
+    # treat all three nodes as an expression when we're in an expression
+    # context
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i, it in n.pairs:
+      result[i] =
+        case it.kind
+        of nkElifExpr, nkElifBranch:
+          parseBranch(c, it, n.kind == nkIfExpr)
+        of nkElseExpr, nkElse:
+          # XXX: hm, also make sure that the 'else' is the last item? Or leave
```

Comment:
Yes, I think this is the type of validation that we _likely_ want to do here.

Although, this level of validation is likely at the edge of complexity we want to entertain, however.

Reply:
The parser also only allows `nkElse`/`nkElseExpr` as the last item, so it makes additional sense to enforce that here too.